Question 5 Part a

Logistic Regression class defined

In [ ]:
class LogisticRegression:

  def __init__(self, learning_rate, iterations=100, w=None, b=0) :
    self.iterations=iterations
    self.learning_rate=learning_rate
    self.b=b
    self.w=w

  def sigmoid(self,x):
    return 1/(1 + np.exp(-x))

  def cross_entropy_loss(self,target,actual):
    return (-target * np.log(actual) - (1 - target) * np.log(1 - actual)).mean() 

  def GetActualValue( self ):     
        return  self.X.dot( self.w) + self.b

  def fit(self,X,Y):
    self.X=X
    self.Y=Y   
    
    self.samples, self.features = X.shape  
    if self.w is None :      
        self.w = np.zeros( self.features) 

    for i in range(self.iterations):
       print('For iteration :')
       print(i)
       self.update_weights()

  def update_weights(self):

    y=self.GetActualValue()
    z=self.sigmoid(y)
    error=z-self.Y.T
    dw=(np.dot(self.X.T,error))/self.samples
    db=np.sum(error)/self.samples

    loss = self.cross_entropy_loss(self.Y.T,z)
    print(f'loss: {loss} \t')  
    print(f'w,b before update: {(self.w,self.b)} \t')

    self.w -= self.learning_rate*dw
    self.b -= self.learning_rate*db

    print(f'w,b after update: {(self.w,self.b)} \t')
        

  def predict( self, X ) : 
        y=np.dot(X,self.w) + self.b
        Z = self.sigmoid(y)    
        Y = np.where( Z > 0.5, 1, 0 )        
        return Y   

Normalization of features

In [ ]:
def FeatureNormalization(features_dataset):
    minmax_col=list()
    num_features=len(features_dataset[0])
    for i in range(num_features):
      values=[row[i] for row in features_dataset]
      min=np.min(values)
      max=np.max(values)
      minmax_col.append([min,max])
    for row in features_dataset:
      for i in range(len(row)):
        row[i]=(row[i]-minmax_col[i][0])/(minmax_col[i][1]-minmax_col[i][0])

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
dataset=pd.read_csv('LR_train_set.csv')

In [ ]:
X_train = dataset.iloc[:,:-1].values
y_train = dataset.iloc[:,-1].values

In [ ]:
dataset_test=pd.read_csv('LR_test_set.csv')

In [ ]:
X_test = dataset_test.iloc[:,:-1].values
y_test = dataset_test.iloc[:,-1].values

In [ ]:
FeatureNormalization(X_train)

In [ ]:
FeatureNormalization(X_test)

Creating model  

In [ ]:
reg=LogisticRegression(0.3,1000)

In [ ]:
reg.fit(X_train,y_train)

For iteration :
0
loss: 0.6931471805599453 	
w,b before update: (array([0., 0.]), 0) 	
w,b after update: (array([0.05112936, 0.03008555]), 0.0) 	
For iteration :
1
loss: 0.6816835250405874 	
w,b before update: (array([0.05112936, 0.03008555]), 0.0) 	
w,b after update: (array([0.10041714, 0.0579665 ]), -0.0029183761961067644) 	
For iteration :
2
loss: 0.6711809496116864 	
w,b before update: (array([0.10041714, 0.0579665 ]), -0.0029183761961067644) 	
w,b after update: (array([0.14804782, 0.08389496]), -0.008376724297595877) 	
For iteration :
3
loss: 0.6614516551227075 	
w,b before update: (array([0.14804782, 0.08389496]), -0.008376724297595877) 	
w,b after update: (array([0.19418594, 0.1080946 ]), -0.01604010889479988) 	
For iteration :
4
loss: 0.6523490933975724 	
w,b before update: (array([0.19418594, 0.1080946 ]), -0.01604010889479988) 	
w,b after update: (array([0.23897741, 0.13076302]), -0.025613263250320484) 	
For iteration :
5
loss: 0.6437591121144365 	
w,b before update: (array([

Predictions on test set

In [ ]:
y_pred=reg.predict(X_test)

In [ ]:
y_pred

array([1, 1, 0, 1, 1, 1])

In [ ]:
y_test

array([0, 0, 0, 1, 1, 1])

Function for creating confusion matrix and returning Accuracy Precision Recall

In [ ]:
def CalAccPrecisionRecall(y_test,y_pred):
  TP=FP=TN=FN=0
  for i in range(len(y_test)):
    if y_pred[i]==y_test[i]:
      if y_test[i]==1:
         TP=TP+1
      else:
         TN=TN+1   
    else:
      if y_test[i]==1:
        FN=FN+1
      else :
        FP=FP+1
   
  return (TP+TN)/(TP+FP+TN+FN),TP/(TP+FP),TP/(TP+FN)     

In [ ]:
CalAccPrecisionRecall(y_test,y_pred)  

(0.6666666666666666, 0.6, 1.0)

Question 5 b part 2 and 3 included

In [ ]:
reg=LogisticRegression(0.1,1000,[1.5,0.5],-1)
reg.fit(X_train,y_train)

For iteration :
0
loss: 0.4390700429767141 	
w,b before update: ([1.5, 0.5], -1) 	
w,b after update: (array([1.51234756, 0.50790003]), -0.9980220558745176) 	
For iteration :
1
loss: 0.4369024973142627 	
w,b before update: (array([1.51234756, 0.50790003]), -0.9980220558745176) 	
w,b after update: (array([1.52453738, 0.51560059]), -0.9963144438046763) 	
For iteration :
2
loss: 0.43481331373336024 	
w,b before update: (array([1.52453738, 0.51560059]), -0.9963144438046763) 	
w,b after update: (array([1.53657517, 0.52310931]), -0.9948660114419572) 	
For iteration :
3
loss: 0.4327969974013927 	
w,b before update: (array([1.53657517, 0.52310931]), -0.9948660114419572) 	
w,b after update: (array([1.54846644, 0.53043352]), -0.9936660591033835) 	
For iteration :
4
loss: 0.43084847400247783 	
w,b before update: (array([1.54846644, 0.53043352]), -0.9936660591033835) 	
w,b after update: (array([1.56021644, 0.53758026]), -0.9927043229091994) 	
For iteration :
5
loss: 0.4289630571862486 	
w,b before 

Question 5b i
What is the logistic model P(ˆy = 1|x1, x2) and its cross-entropy error
function?

As we know the logistic model for P(ˆy = 1|x1, x2) by substituiting θ,θ1,θ2 given as -1,1.5,0.5

**fθ(x1, x2)=1.5* x1 + 0.5* x2 -1**

Cross Entropy Error Function is

**loss=(-target * np.log(actual) - (1 - target) * np.log(1 - actual)).mean() **
where target is the desired value and actual is the value from model i.e. yhat
 

Question 5b (ii)

As we can see after one iteration values of W=[1.51234756, 0.50790003]) and b= -0.9980220558745176

Hence model before first iteration is

 **fθ(x1, x2)=1.5* x1 + 0.5* x2 -1**

After first iteration it is

**fθ(x1, x2)=1.51234756 * x1 + 0.50790003 * x2 - 0.9980220558745176**


In [ ]:
y_pred=reg.predict(X_test)

Question 5 b (iii)

In [ ]:
Acc,precision,recall=CalAccPrecisionRecall(y_test,y_pred)
print("Accuracy :")
print(Acc)
print("Precision :")
print(precision)
print("Recall :")
print(recall)

Accuracy :
0.6666666666666666
Precision :
0.6
Recall :
1.0
